### PR TITLE
feat(avoidance): use intersection area

### PR DESCRIPTION
**Pull request by satoshi-ota**
### <samp>🤖 Generated by Copilot at 48eabc2</samp>

This pull request refactors and improves the code for generating and using the drivable area in the behavior path planner. It removes redundant code, adds new parameters and flags, and simplifies the logic for expanding the drivable area by polygons such as hatched road markings and intersection areas. It also makes the functions for drivable area calculation more consistent and flexible. The changes affect several files in the `planning/behavior_path_planner` directory, such as `utils.hpp`, `utils.cpp`, `planner_manager.cpp`, and `avoidance.param.yaml`.

Please approve :arrow_down: before this PR.
https://github.com/autowarefoundation/autoware_launch/pull/439

image

image

## Tests performed

Psim

image

image

- [x] [PASS TIER IV INTERNAL SCENARIOS](https://evaluation.tier4.jp/evaluation/reports/a0b93c42-a4d2-5086-add3-44b919392a24?project_id=prd_jt)

## Effects on system behavior

Not applicable.

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: planning/behavior_path_planner/src/utils/utils.cpp [hunk 1]
```diff
@@ -1573,60 +1597,99 @@ std::vector<geometry_msgs::msg::Point> calcBound(
...
+    const auto bound_lane = is_left ? drivable_lane.left_lane : drivable_lane.right_lane;
+    const auto bound = is_left ? bound_lane.leftBound3d() : bound_lane.rightBound3d();
+
+    if (bound.size() < 2) {
+      continue;
+    }
+
+    // expand drivable area by intersection areas.
+    const std::string id = bound_lane.attributeOr("intersection_area", "else");
+    if (enable_expanding_intersection_areas && id != "else") {
```
Comment (kyoichi-sugahara) [inline planning/behavior_path_planner/src/utils/utils.cpp:1611]:
```
    const bool isIntersectionArea = enable_expanding_intersection_areas && intersection_area != "else";

    // expand drivable area by intersection areas.
    if (isIntersectionArea) {
```

Comment (kyoichi-sugahara) [inline planning/behavior_path_planner/src/utils/utils.cpp:1611]:
```
    const bool isIntersectionArea = enable_expanding_intersection_areas && intersection_area != "else";

    // expand drivable area by intersection areas.
    if (isIntersectionArea) {
```

File: planning/behavior_path_planner/src/utils/avoidance/utils.cpp [hunk 2]
```diff
@@ -1137,29 +1137,48 @@ void filterTargetObjects(
...
-      expandable_polygons.push_back(*new_polygon_candidate);
+    if (area_id_str != "else") {
+      expandable_polygons.push_back(
+        rh->getLaneletMapPtr()->polygonLayer.get(std::atoi(area_id_str.c_str())));
     }
   }
 
+  if (expandable_polygons.empty()) {
+    return to_road_shoulder_distance;
+  }
```
Comment (kyoichi-sugahara) [inline planning/behavior_path_planner/src/utils/avoidance/utils.cpp:1180]:
```
  auto polygonExistsIn = [](const auto &polygons, const auto &new_polygon) {
    return std::any_of(polygons.begin(), polygons.end(), [&](const auto &polygon){
      return polygon.id() == new_polygon.id();
    });
  };

  if (use_hatched_road_markings) {
    for (const auto & point : target_line) {
      const auto new_polygon_candidate = utils::getPolygonByPoint(rh, point, "hatched_road_markings");

      if (new_polygon_candidate && !polygonExistsIn(expandable_polygons, *new_polygon_candidate)) {
        expandable_polygons.push_back(*new_polygon_candidate);
      }
    }
  }

  if (use_intersection_areas) {
    const std::string area_id_str = overhang_lanelet.attributeOr("intersection_area", "else");

    if (area_id_str != "else") {
      expandable_polygons.push_back(
        rh->getLaneletMapPtr()->polygonLayer.get(std::atoi(area_id_str.c_str())));
    }
  }

  if (expandable_polygons.empty()) {
    return to_road_shoulder_distance;
  }
```

**Review by kyoichi-sugahara — APPROVED** [review]
LGTM
(suggest a minor modification but it's not necessarly condidered.)

File: planning/behavior_path_planner/src/utils/utils.cpp (hunk 1 above)
Comment (satoshi-ota) [reply planning/behavior_path_planner/src/utils/utils.cpp]: Thanks for your advice! I fixed in https://github.com/autowarefoundation/autoware.universe/pull/4206/commits/1e60bcb87a55dab28ac89e8486de728dd2bc6f6c.

File: planning/behavior_path_planner/src/utils/avoidance/utils.cpp (hunk 2 above)
Comment (satoshi-ota) [reply planning/behavior_path_planner/src/utils/avoidance/utils.cpp]: I fixed in https://github.com/autowarefoundation/autoware.universe/pull/4206/commits/1e60bcb87a55dab28ac89e8486de728dd2bc6f6c.

**Review by takayuki5168 — APPROVED** [review]

**Pull request merged**